src/type-data-frame.c

```diff
@@ -437,8 +437,7 @@ SEXP df_cast(SEXP x, SEXP to, struct vctrs_arg* x_arg, struct vctrs_arg* to_arg)
   // Restore data frame size before calling `vec_restore()`. `x` and
   // `to` might not have any columns to compute the original size.
   init_data_frame(out, size);
-
-  out = PROTECT(vec_restore(out, to, R_NilValue));
+  Rf_setAttrib(out, R_RowNamesSymbol, df_rownames(x));
```

Comment:
This is another example where casting from a data frame with row names to tibble would carry the row names onto the tibble, but as you said about `vec_set_names()`, I think that would be okay, especially since tibble allows the row names, but hides them. (Oh I see you have a new test for this, 👍 )